### ISSUE TITLE
Multiple RTSPs with specific YOLOv8 models in each stream

### ISSUE THREAD

**Issue by pornpra** [ultralytics/ultralytics#7062]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, Suppose I have
1. Pretrained YOLOv8 (model 1)
2. Custom YOLOv8 (model 2)
3. Custom YOLOv8 (model 3)

**How can I use Ultralytics to run multiple RTSPs with specific model in each stream** e.g.
- RTSP 1 inferences with model 1
- RTSP 2 inferences with model 2
- RTSP 3 inferences with model 3
- RTSP 4 inferences with model 1

My current environement: Python 3.8.10

Thanks

**Comment by reno77** [comment]
Simple way ... from phind.com 

```
import cv2
from ultralytics import YOLO
import threading

model1 = YOLO('path_to_model1')
model2 = YOLO('path_to_model2')
model3 = YOLO('path_to_model3')

rtsp_streams = ['rtsp://stream1', 'rtsp://stream2', 'rtsp://stream3', 'rtsp://stream4']

def run_inference_on_stream(stream, model):
   video = cv2.VideoCapture(stream)

   while True:
       ret, frame = video.read()

       if not ret:
           break

       results = model.track(frame, persist=True)
       res_plotted = results[0].plot()
       cv2.imshow("Inference Stream", res_plotted)

       key = cv2.waitKey(1)
       if key == ord('q'):
           break

   video.release()
 

for i, stream in enumerate(rtsp_streams):
   model = [model1, model2, model3][i % 3] # cycle through the models
   thread = threading.Thread(target=run_inference_on_stream, args=(stream, model))
   thread.start()
```

**Comment by glenn-jocher** [comment]
@reno77 it looks like you're looking for a way to perform real-time inference with YOLOv8 on multiple RTSP streams, each using a different YOLOv8 model. While I can't provide a full code example, I can certainly point you in the right direction.

You'll want to create separate threads or processes for each RTSP stream to avoid blocking. Each of these threads will handle the inference using one of your models (model 1, model 2, model 3). To implement this, initialize each of your models separately, create a function that captures frames from the RTSP stream, performs inference using the designated model, and processes the results as needed.

When iterating over your RTSP streams, assign the appropriate model to each stream and invoke your inference function in a new thread. The threading module will be useful for this. Make sure to handle thread synchronization and termination conditions appropriately.

Consider using the `Predict` mode for straightforward inference or the `Track` mode if you're interested in tracking objects over time.

For reference on the modes I mentioned and for additional guidance on working with the Ultralytics YOLOv8 models, please refer to our Ultralytics Docs.

Good luck with your multi-stream real-time inference project! 😊🚀

**Comment by glenn-jocher** [comment]
The error you’re encountering, `[h264 @ ...] error while decoding MB ...`, is likely related to either RTSP stream decoding (not directly GPU-related) or system resource limitations for video processing and inference. Your current system's CPU could be the bottleneck, especially without GPU acceleration for inference tasks.

To address the issue:
1. **RTSP Handling**: Ensure your `ffmpeg` and OpenCV installations are up-to-date, as older versions can cause decoding errors. Additionally, confirm that your RTSP stream is stable and does not have packet loss.
2. **CPU Bottleneck**: Video decoding and YOLO inference are resource-intensive, so relying solely on your CPU may not be efficient. A mid-range GPU is highly recommended.

### Suitable GPUs for YOLOv8
- **Cost-Effective Options**: NVIDIA RTX 3060 or RTX 4060 - Both offer good price-performance ratios and adequate VRAM (12GB for the RTX 3060) for most YOLOv8 tasks.
- **Higher-Performance Options**: NVIDIA RTX 4070 or RTX 4080 – Suitable for heavier workloads with larger datasets or higher resolution inputs.
- **Budget Options**: If constrained by budget, GTX 1660 Super may also work for training smaller models though may struggle with higher demands.

For inference needs rather than training, even lower-tier GPUs like the RTX 3050 can suffice. Always consider the VRAM requirements of your model resolution and batch size. If cost is a key factor, RTX 3060 offers a well-rounded balance.

For more optimized GPU recommendations, check benchmarks or engage with the [YOLO community](https://github.com/ultralytics/ultralytics/discussions) for firsthand insights!

**Comment by akdenmikail** [comment]
@glenn-jocher Thank you very much for this valuable information.:)

**Comment by glenn-jocher** [comment]
You're very welcome! We're glad you found it helpful. All credit goes to the YOLO community and the Ultralytics team for their collaborative efforts. Let us know if you have any further questions!
